### PR TITLE
Rewrite patternDB test cases based on criterion

### PR DESCRIPTION
**Pull request by norberttak**
Rewrite patternDB test cases based on criterion

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/dbparser/tests/test_patterndb.c [hunk 1]
```diff
@@ -706,434 +441,223 @@ test_simple_rule_with_rate_limited_action(void)
...
-  g_file_open_tmp("patterndbXXXXXX.xml", &filename, NULL);
-  g_file_set_contents(filename, tag_outside_of_rule_skeleton,
-                      strlen(tag_outside_of_rule_skeleton), NULL);
+ParameterizedTestParameters(pattern_db, test_rules)
+{
+  static PatternDBTestParam parser_params[] =
+  {
+    // test_conflicting_rules_with_different_parsers
+    {
+      .pattern_db_file = PATH_PREFIX"pdb_conflicting_rules_with_different_parsers.xml",
```
Comment (Kokan) [inline modules/dbparser/tests/test_patterndb.c]: I think this points toward a wrong direction. The unit test should not have external files that they are depending on. 
I know that the existing version created and loaded that file (but I think that was less evil.)

This is probably out of current PR's scope, just to give you a hint, where I think it should move and why is it important not to imply the current is the good direction.

The `PatternDB` itself should not know where its contents come: file, user input, stream or whatever. Try to think how you would modify the `PatternDB` if you would want to add a feature like embedded `PatternDB` in syslog-ng configuration (like `python`).

```
patterndb inlinepattern {
 <patterndb version='5' pub_date='2010-02-22'>
 ... 
</patterndb>
};
```

Actually this is exactly what we would like to achieve in the unit test IMHO.

(Anyway if this is kept as it is - :disappointed: , those files should be added to the `EXTRA_DIST` target.)

File: modules/dbparser/tests/CMakeLists.txt [hunk 2]
```diff
@@ -1,6 +1,7 @@
 add_unit_test(TARGET test_timer_wheel DEPENDS patterndb)
 add_unit_test(TARGET test_patternize DEPENDS patterndb syslogformat)
-add_unit_test(CRITERION LIBTEST TARGET test_patterndb DEPENDS patterndb basicfuncs syslogformat)
+add_unit_test(LIBTEST TARGET test_patterndb DEPENDS patterndb basicfuncs syslogformat)
```
Comment (Kokan) [inline modules/dbparser/tests/CMakeLists.txt]: This patch reverts this line, this line should not be changed in this patch.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/dbparser/tests/test_parsers.c [hunk 3]
```diff
@@ -74,56 +56,77 @@ _invoke_parser(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gp
...
-  gboolean result;
+typedef gboolean (*ParserFunc)(guint8 *str, gint *len, const gchar *param, gpointer state, RParserMatch *match);
 
-  result = _invoke_parser(parser, str, param, state, &result_string);
-  assert_true(result, "Mismatching parser result");
-  assert_string(result_string, expected_string, "Mismatching parser result");
-  g_free(result_string);
+typedef struct _parser_test_param
+{
+  ParserFunc parser;
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/dbparser/tests/test_parsers.c]: OK, removed the parser member from structure

File: modules/dbparser/tests/test_parsers.c [hunk 4]
```diff
@@ -136,35 +139,16 @@ compile_qstring_state(const gchar *quotes)
...
-
-void
-test_qstring(void)
+ParameterizedTest(ParserQStringTestParam *param, parser, test_qstring_parser)
 {
-  PARSER_TESTCASE(test_qstring_parser_extracts_word_from_quotes);
-}
-
-/* NOTE: incomplete, the rest of the testcases are missing */
+//  assert_parser_success(r_parser_qstring, str, quotes, _compile_qstring_state(quotes), expected_string);
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/dbparser/tests/test_parsers.c]: removed

File: modules/dbparser/tests/test_parsers.c [hunk 5]
```diff
@@ -74,56 +56,77 @@ _invoke_parser(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gp
...
-  result = _invoke_parser(parser, str, param, state, &result_string);
-  assert_true(result, "Mismatching parser result");
-  assert_string(result_string, expected_string, "Mismatching parser result");
-  g_free(result_string);
+typedef struct _parser_test_param
+{
+  ParserFunc parser;
+  gchar *str;
+  gpointer param;
+  gpointer state;
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/dbparser/tests/test_parsers.c]: removed

File: modules/dbparser/tests/test_patterndb.c (hunk 1 above)
Comment (norberttak) [reply modules/dbparser/tests/test_patterndb.c]: I don't really catch the point. I need some more details about your concern. let's discuss it personally...

Comment (norberttak) [reply modules/dbparser/tests/test_patterndb.c]: I removed the XML files and moved their content into test_patterndb.h header file.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by norberttak** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/dbparser/tests/test_parsers.c [hunk 6]
```diff
@@ -74,56 +56,72 @@ _invoke_parser(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gp
...
+    {.str = "{foo}", .quotes = "{}", .expected_string = "foo"},
+  };
 
-void
-test_string(void)
-{
-  PARSER_TESTCASE(test_string_parser_without_parameter_parses_a_word);
-  PARSER_TESTCASE(test_string_parser_with_additional_end_characters);
+  return cr_make_param_array(ParserQStringTestParam, parser_params,
+                             sizeof(parser_params) / sizeof(parser_params[0]));
```
Comment (furiel) [inline modules/dbparser/tests/test_parsers.c]: Can I suggest https://developer.gnome.org/glib/stable/glib-Standard-Macros.html#G-N-ELEMENTS:CAPS?

File: modules/dbparser/tests/test_parsers.c [hunk 7]
```diff
@@ -74,56 +56,72 @@ _invoke_parser(gboolean (*parser)(guint8 *str, gint *len, const gchar *param, gp
...
+  gchar *result_string = NULL;
   gboolean result;
 
-  result = _invoke_parser(parser, str, param, state, NULL);
-  assert_false(result, "Mismatching parser result");
+  result = _invoke_parser(r_parser_string, param->str, param->param, NULL, &result_string);
+  if (param->expected_result == TRUE)
+    {
+      cr_assert(result, "Mismatching parser result");
+      cr_assert_str_eq(result_string, param->expected_string, "Mismatching parser result");
```
Comment (furiel) [inline modules/dbparser/tests/test_parsers.c]: I'm afraid this hides the strings in case of error. We should either not provide an error string, or print the underlying strings similarly to the bottom of the file as %s %s.

File: modules/dbparser/tests/test_patterndb.h [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+              </tags>\
+            </message>\
+          </action>\
+        </actions>\
+     </rule>\
+    </rules>\
+  </ruleset>\
+</patterndb>"
+
+#define pdb_msg_count_skeleton "<patterndb version='4' pub_date='2010-02-22'>\
```
Comment (furiel) [inline modules/dbparser/tests/test_patterndb.h]: If I understand correctly, these were originally gchar * types. Why is these changed to defines?

File: modules/dbparser/tests/CMakeLists.txt (hunk 2 above)
Comment (Kokan) [reply modules/dbparser/tests/CMakeLists.txt]: The 2nd patch modified the `add_unit_test(... test_patterndb ...)`, so it is `add_unit_test(CRITERION LIBTEST TARGET test_patterndb DEPENDS patterndb basicfuncs syslogformat)` with Criterion.
And the last (3rd patch), reverts the above change, so the result is the `test_patterndb` is marked as non-criterion test.
(In the unified diff it cannot be seen.)

File: modules/dbparser/tests/test_parsers.c (hunk 7 above)
Comment (norberttak) [reply modules/dbparser/tests/test_parsers.c]: I added the result and expected strings to printout.

File: modules/dbparser/tests/test_patterndb.h (hunk 8 above)
Comment (norberttak) [reply modules/dbparser/tests/test_patterndb.h]: I made the XML strings as a macro because I don't like to create variables in header files. If we want to use the const gchar*  then we should create C file instead of H.
I'd like to keep it as is.

File: modules/dbparser/tests/test_parsers.c (hunk 6 above)
Comment (norberttak) [reply modules/dbparser/tests/test_parsers.c]: I added the G_N_ELEMENTS macro.

File: modules/dbparser/tests/CMakeLists.txt (hunk 2 above)
Comment (norberttak) [reply modules/dbparser/tests/CMakeLists.txt]: Thanks for the finding. I clean up the commits.

**Comment by kira-syslogng** [comment]
Build SUCCESS
